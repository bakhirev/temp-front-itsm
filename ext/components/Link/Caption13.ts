import styled from 'styled-components';

import { Body2Long } from '../Typography';

export const Caption13 = styled(Body2Long)`
  white-space: pre-wrap;
  color: inherit;
`;
