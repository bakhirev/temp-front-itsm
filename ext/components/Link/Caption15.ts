import styled from 'styled-components';

import { Body1Long } from '../Typography';

export const Caption15 = styled(Body1Long)`
  white-space: pre-wrap;
  color: inherit;
`;
