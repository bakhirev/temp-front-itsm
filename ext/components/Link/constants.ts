export type LinkSize = 'big' | 'small';
export type LinkKind = 'primary' | 'secondary';
export type LinkIconPosition = 'left' | 'right';
export type LinkTarget = '_self' | '_parent' | '_top' | '_blank';

export const LINK_ICON_PADDING = '6px';

export const SIZES = {
  big: 24,
  small: 20,
};
