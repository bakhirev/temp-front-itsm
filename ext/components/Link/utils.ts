import type { ITheme } from '../themes';

import type { LinkKind } from './constants';

export const getColors = (theme: ITheme, kind: LinkKind, inverse: boolean) => {
  if (inverse) {
    const inverseColor = theme.color.primary[40];
    return {
      color: inverseColor,
      fill: inverseColor,
      hover: {
        color: inverseColor,
        fill: inverseColor,
      },
      focus: {
        color: inverseColor,
        fill: inverseColor,
      },
    };
  }
  switch (kind) {
    case 'primary':
      return {
        color: theme.color.primary[60],
        fill: theme.color.primary[60],
        hover: {
          color: theme.color.primary[70],
          fill: theme.color.primary[70],
        },
        focus: {
          color: theme.color.primary[60],
          fill: theme.color.primary[60],
        },
      };
    case 'secondary':
      return {
        color: theme.color.neutral[90],
        fill: theme.color.neutral[50],
        hover: {
          color: theme.color.primary[60],
          fill: theme.color.primary[60],
        },
        focus: {
          color: theme.color.neutral[90],
          fill: theme.color.neutral[50],
        },
      };
  }
};
