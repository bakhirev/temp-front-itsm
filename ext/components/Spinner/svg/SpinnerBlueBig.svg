<svg width="48" height="48" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M43.9553 37.3337C46.0346 34.2217 47.3495 30.6787 47.8128 26.9922C48.0193 25.3483 46.6569 24 45 24C43.3432 24 42.0262 25.3532 41.7532 26.9874C40.3302 35.5067 32.9234 42 24 42C14.0589 42 6.00001 33.9411 6.00001 24C6.00001 15.0766 12.4933 7.6698 21.0126 6.24674C22.6468 5.97377 24 4.65683 24 2.99997C24 1.34312 22.6517 -0.0193537 21.0077 0.187219C17.3213 0.650452 13.7783 1.96536 10.6663 4.0447C6.71955 6.68186 3.64341 10.4301 1.8269 14.8156C0.0103989 19.201 -0.464881 24.0266 0.461164 28.6821C1.38721 33.3377 3.67299 37.6141 7.02945 40.9705C10.3859 44.327 14.6623 46.6128 19.3178 47.5388C23.9734 48.4649 28.799 47.9896 33.1844 46.1731C37.5698 44.3566 41.3181 41.2804 43.9553 37.3337Z" fill="#0062FF"/>
</svg>
