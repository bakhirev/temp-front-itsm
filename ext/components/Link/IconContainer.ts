import styled from 'styled-components';

import { SIZES } from './constants';
import type { LinkSize } from './constants';

export const IconContainer = styled.div<{ size: LinkSize }>`
  display: inline-block;
  vertical-align: bottom;
  ${({ size }) => `
    width: ${SIZES[size]}px;
    height: ${SIZES[size]}px;
  `}
`;
