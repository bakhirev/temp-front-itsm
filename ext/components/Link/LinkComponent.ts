import styled from 'styled-components';

import { DEFAULT_THEME } from '../common';

import { Caption13 } from './Caption13';
import { Caption15 } from './Caption15';
import { SIZES, LINK_ICON_PADDING } from './constants';
import { getColors } from './utils';
import type { LinkKind, LinkSize, LinkIconPosition } from './constants';

export type BasicLinkProps = React.AnchorHTMLAttributes<HTMLAnchorElement>;

interface ILinkComponentProps extends BasicLinkProps {
  kind: LinkKind;
  size: LinkSize;
  iconPosition?: LinkIconPosition;
  inverse: boolean;
  disabled: boolean;
}

export const LinkComponent = styled.a<ILinkComponentProps>`
  outline: none;
  position: relative;
  display: inline;
  box-sizing: border-box;
  ${({ size }) => `
    height: ${SIZES[size]}px;
  `}

  ${({ theme, kind, iconPosition, inverse, disabled }) =>
    disabled
      ? `
      pointer-events: none;
      cursor: default;
      color: ${inverse ? theme.color.neutral[50] : theme.color.neutral[30]};
      svg {
        fill: ${inverse ? theme.color.neutral[50] : theme.color.neutral[30]};
      }
    `
      : `
      cursor: pointer;
      color: ${getColors(theme, kind, inverse).color};
      svg {
        fill: ${getColors(theme, kind, inverse).fill};
      }

      &:hover, &:active {
        color: ${getColors(theme, kind, inverse).hover.color};
        ${
          !iconPosition
            ? `border-bottom: 1px solid ${getColors(theme, kind, inverse).hover.color};`
            : ''
        }
        svg {
          fill: ${getColors(theme, kind, inverse).hover.fill};
        }
      }

      &:focus {
        color: ${getColors(theme, kind, inverse).focus.color};
        border-bottom: 2px solid ${inverse ? theme.color.primary[40] : theme.color.primary[60]};
        svg {
          fill: ${getColors(theme, kind, inverse).focus.fill};
        }
      }
  `}

  ${Caption13}, ${Caption15} {
    display: inline;
    ${({ iconPosition }) => `
      padding-left: ${iconPosition === 'left' ? LINK_ICON_PADDING : 0};
      padding-right: ${iconPosition === 'right' ? LINK_ICON_PADDING : 0};
    `}
  }
`;

LinkComponent.defaultProps = {
  theme: DEFAULT_THEME,
};
