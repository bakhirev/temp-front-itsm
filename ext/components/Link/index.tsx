import React, { ReactNode, ElementType, FC } from 'react';

import { LinkComponent } from './LinkComponent';
import { Caption13 } from './Caption13';
import { Caption15 } from './Caption15';
import { IconContainer } from './IconContainer';
import { SIZES } from './constants';
import type { LinkKind, LinkSize, LinkIconPosition, LinkTarget } from './constants';
import type { BasicLinkProps } from './LinkComponent';

export interface ILinkProps extends BasicLinkProps {
  /** Текст ссылки */
  children: ReactNode;
  /** Гиперссылка, представляет URL или якорь (#ID элемента на текущей странице) */
  href: string;
  /** Иконка, отображаемая внутри ссылки слева или справа */
  icon?: ElementType;
  /** Вид ссылки */
  kind?: LinkKind;
  /** Размер ссылки */
  size?: LinkSize;
  /** Размещение иконки слева или справа текста ссылки */
  iconPosition?: LinkIconPosition;
  /* Где отобразить переход URL */
  target?: LinkTarget;
  /** Инвертировать цвет ссылки для отображения на тёмных фонах */
  inverse?: boolean;
  /** Отключение ссылки */
  disabled?: boolean;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
}

export const Link: FC<ILinkProps> = ({
  children,
  href,
  icon: Icon,
  kind = 'primary',
  size = 'big',
  target,
  iconPosition = Icon ? 'right' : undefined,
  inverse = false,
  disabled = false,
  className,
  dataTestId,
  ...props
}: ILinkProps) => {
  const Caption = size === 'big' ? Caption15 : Caption13;
  const iconSize = SIZES[size];
  return (
    <LinkComponent
      {...props}
      href={href}
      target={target}
      inverse={inverse}
      className={className}
      data-test-id={dataTestId}
      kind={kind}
      size={size}
      iconPosition={Icon ? iconPosition : undefined}
      disabled={disabled}
    >
      {Icon && iconPosition === 'left' && (
        <IconContainer size={size}>{<Icon width={iconSize} height={iconSize} />}</IconContainer>
      )}
      <Caption>{children}</Caption>
      {Icon && iconPosition === 'right' && (
        <IconContainer size={size}>{<Icon width={iconSize} height={iconSize} />}</IconContainer>
      )}
    </LinkComponent>
  );
};
