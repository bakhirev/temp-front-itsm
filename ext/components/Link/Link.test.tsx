import React from 'react';
import { shallow } from 'enzyme';

import { ReactComponent as ArrowLeftOutline } from '../Icons/system/ArrowLeftOutline.svg';
import { Link } from '../Link';

const linkText = 'Link Text';

jest.mock('../common', () => ({
  DEFAULT_THEME: {},
}));

describe('Link', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render component by default as "PRIMARY" link', () => {
    const wrapper = shallow(<Link href="#0">{linkText}</Link>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind="SECONDARY" link', () => {
    const wrapper = shallow(
      <Link href="#0" kind={'secondary'}>
        {linkText}
      </Link>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with size="SMALL" link', () => {
    const wrapper = shallow(
      <Link href="#0" size={'small'}>
        {linkText}
      </Link>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind="PRIMARY" link disabled', () => {
    const wrapper = shallow(
      <Link href="#0" kind={'primary'} disabled>
        {linkText}
      </Link>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component by default link with icon without icon position, icon position RIGHT by default', () => {
    const wrapper = shallow(
      <Link href="#0" icon={ArrowLeftOutline}>
        {linkText}
      </Link>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component by default link with icon and icon position LEFT', () => {
    const wrapper = shallow(
      <Link href="#0" icon={ArrowLeftOutline} iconPosition={'left'}>
        {linkText}
      </Link>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component by default link with icon and icon position RIGHT', () => {
    const wrapper = shallow(
      <Link href="#0" icon={ArrowLeftOutline} iconPosition={'right'}>
        {linkText}
      </Link>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component by default as "PRIMARY" link with inverse option', () => {
    const wrapper = shallow(
      <Link href="#0" inverse>
        {linkText}
      </Link>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
