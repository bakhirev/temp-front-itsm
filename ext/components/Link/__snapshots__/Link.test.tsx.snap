// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render component by default as "PRIMARY" link 1`] = `
<styled.a
  disabled={false}
  href="#0"
  inverse={false}
  kind="primary"
  size="big"
  theme={Object {}}
>
  <Styled(styled.div)>
    Link Text
  </Styled(styled.div)>
</styled.a>
`;

exports[`Link should render component by default as "PRIMARY" link with inverse option 1`] = `
<styled.a
  disabled={false}
  href="#0"
  inverse={true}
  kind="primary"
  size="big"
  theme={Object {}}
>
  <Styled(styled.div)>
    Link Text
  </Styled(styled.div)>
</styled.a>
`;

exports[`Link should render component by default link with icon and icon position LEFT 1`] = `
<styled.a
  disabled={false}
  href="#0"
  iconPosition="left"
  inverse={false}
  kind="primary"
  size="big"
  theme={Object {}}
>
  <styled.div
    size="big"
  >
    <div
      height={24}
      width={24}
    />
  </styled.div>
  <Styled(styled.div)>
    Link Text
  </Styled(styled.div)>
</styled.a>
`;

exports[`Link should render component by default link with icon and icon position RIGHT 1`] = `
<styled.a
  disabled={false}
  href="#0"
  iconPosition="right"
  inverse={false}
  kind="primary"
  size="big"
  theme={Object {}}
>
  <Styled(styled.div)>
    Link Text
  </Styled(styled.div)>
  <styled.div
    size="big"
  >
    <div
      height={24}
      width={24}
    />
  </styled.div>
</styled.a>
`;

exports[`Link should render component by default link with icon without icon position, icon position RIGHT by default 1`] = `
<styled.a
  disabled={false}
  href="#0"
  iconPosition="right"
  inverse={false}
  kind="primary"
  size="big"
  theme={Object {}}
>
  <Styled(styled.div)>
    Link Text
  </Styled(styled.div)>
  <styled.div
    size="big"
  >
    <div
      height={24}
      width={24}
    />
  </styled.div>
</styled.a>
`;

exports[`Link should render component with kind="PRIMARY" link disabled 1`] = `
<styled.a
  disabled={true}
  href="#0"
  inverse={false}
  kind="primary"
  size="big"
  theme={Object {}}
>
  <Styled(styled.div)>
    Link Text
  </Styled(styled.div)>
</styled.a>
`;

exports[`Link should render component with kind="SECONDARY" link 1`] = `
<styled.a
  disabled={false}
  href="#0"
  inverse={false}
  kind="secondary"
  size="big"
  theme={Object {}}
>
  <Styled(styled.div)>
    Link Text
  </Styled(styled.div)>
</styled.a>
`;

exports[`Link should render component with size="SMALL" link 1`] = `
<styled.a
  disabled={false}
  href="#0"
  inverse={false}
  kind="primary"
  size="small"
  theme={Object {}}
>
  <Styled(styled.div)>
    Link Text
  </Styled(styled.div)>
</styled.a>
`;
